src/client/client_impl.cc

```diff
@@ -79,6 +82,189 @@ google::protobuf::util::Status ClientImpl::CreateFile(
+      chunk_version = chunk_version_or.ValueOrDie(); 
+      auto chunk_server_location_entry_or(
+          cache_manager_->GetChunkServerLocation(chunk_handle));
+      if (chunk_server_location_entry_or.ok()) {
+        chunk_server_location_entry = chunk_server_location_entry_or
+                                          .ValueOrDie();
+      } else {
+        file_chunk_metadata_in_cache = false;
+      }
+    } else {
+      file_chunk_metadata_in_cache = false;
+    }
+  } else {
+    file_chunk_metadata_in_cache = false;
+  }
+
+  // If file chunk metadata is not available, issue an OpenFileRequest to get
+  // all information needed
+  if (!file_chunk_metadata_in_cache) {
+    // TODO(Xi): refactor this block into a helper function
+    // Prepare the open file request
+    OpenFileRequest open_file_request;
+    open_file_request.set_filename(filename);
+    open_file_request.set_chunk_index(chunk_index);
+    open_file_request.set_mode(OpenFileRequest::READ);
+    grpc::ClientContext client_context;
+    common::SetClientContextDeadline(client_context, config_manager_);
+
+    LOG(INFO) << "Issuing OpenFilRequest to read file " << filename
+              << " at chunk index " << chunk_index;
+
+    // Issue OpenFileReply rpc and check stastus
+    StatusOr<OpenFileReply> open_file_or(
+          master_metadata_service_client_->SendRequest(
+              open_file_request, client_context));
+
+    // Handle error
+    if (!open_file_or.ok()) {
+      LOG(ERROR) << "OpenFileRequest failed due to " 
+                 << open_file_or.status().error_message();
+      return open_file_or.status();
+    }
+
+    // Cache file chunk metadata
+    auto open_file_reply(open_file_or.ValueOrDie());
+    cache_file_chunk_metadata(filename, chunk_index, open_file_reply);
+  
+    chunk_handle = cache_manager_->GetChunkHandle(filename, chunk_index)
+                                       .ValueOrDie();
+    chunk_version = cache_manager_->GetChunkVersion(chunk_handle).ValueOrDie();
+    chunk_server_location_entry = 
+        cache_manager_->GetChunkServerLocation(chunk_handle).ValueOrDie();
+  }
+
+  // If no locations are there, we need to return error
+  if (chunk_server_location_entry.locations.empty()) {
+    return google::protobuf::util::Status(
+               google::protobuf::util::error::UNAVAILABLE,
+               "No chunk server has been found");
+  }
+
+  // Loop over all server locations but return immediately as long as one 
+  // read request is successful
+  for (auto& location : chunk_server_location_entry.locations) {
+    // Prepare for the read file chunk request
+    ReadFileChunkRequest read_file_chunk_request;
+    read_file_chunk_request.set_chunk_handle(chunk_handle);
+    read_file_chunk_request.set_chunk_version(chunk_version);
+    // We start reading from the beginning of this chunk, where the offset
+    // equals to chunk_index * common::bytes_per_mb
+    read_file_chunk_request.set_offset_start(chunk_index * 
```

Comment:
this is def not right. the offset_start should be the client's requested "offset_start" divided by "GFS block size", and take the remainder to see the offset_start for the *first block*, and offset_start = 0 for subsequent blocks. 

However, since this is ReadFileChunk, this "offset" should be set by a parameter passed to the function, and that parameter is in turn determined by the "ReadFile" implementation

Reply:
see my concrete comment later in the PR review for ReadFile for better instructions

Reply:
Apparently my brain is not working when I wrote this logic, an offset is de needed here, see more response below.